### ISSUE TITLE
Add KServe Model Proxy Integration (ADR-039 Phase 4)

### ISSUE THREAD

**Issue by tosin2013** [tosin2013/openshift-coordination-engine#18]
## Overview

Implement KServe model proxy integration in the Go coordination engine to enable notebooks to call KServe InferenceServices through the coordination engine as a central orchestrator.

## Problem Statement

Notebooks currently have no standardized way to call KServe models through the coordination engine. Users need to either:
- Call KServe InferenceServices directly (duplicates configuration)
- Embed ML models in notebooks (hard to fork/modify)

Per **ADR-039** (User-Deployed KServe Models) and **ADR-040** (Extensible KServe Model Registry), the coordination engine should act as a proxy to KServe models.

## Architecture Goal

```
Notebooks → Coordination Engine → KServe InferenceServices
            /api/v1/detect        (user-deployed models)
```

**Benefits**:
- ✅ Coordination engine as central orchestrator
- ✅ Users fork notebooks easily (config-driven)
- ✅ Supports extensible model registry
- ✅ GitOps-native workflow

### 1. Read KServe Configuration from Environment Variables

The deployment already provides these environment variables (from `values-hub.yaml`):

```yaml
env:
  - name: ENABLE_KSERVE_INTEGRATION
    value: "true"
  - name: KSERVE_NAMESPACE
    value: "self-healing-platform"
  - name: KSERVE_ANOMALY_DETECTOR_SERVICE
    value: "anomaly-detector-predictor"
  - name: KSERVE_PREDICTIVE_ANALYTICS_SERVICE
    value: "predictive-analytics-predictor"
```

### 2. Implement `/api/v1/detect` Endpoint

**HTTP Method**: POST

**Request Format**:
```json
{
  "model": "anomaly-detector",
  "instances": [
    [0.5, 1.2, 0.8],
    [0.3, 0.9, 1.1]
  ]
}
```

**Response Format** (KServe v1 protocol):
```json
{
  "predictions": [-1, 1],
  "model_name": "anomaly-detector",
  "model_version": "v2"
}
```

**Error Handling**:
- Return 404 if model not registered
- Return 503 if KServe model unavailable
- Return 400 if invalid request format

### 3. Support Dynamic Model Discovery (ADR-040)

The coordination engine should support **any model** registered in `values-hub.yaml` without code changes.

**Current** (hardcoded - ❌):
```go
models := map[string]string{
    "anomaly-detector": "anomaly-detector-predictor",
    "predictive-analytics": "predictive-analytics-predictor",
}
```

**Required** (dynamic from env vars - ✅):
```go
func loadModelsFromEnv() map[string]string {
    models := make(map[string]string)

    // Load models from environment variables
    // Pattern: KSERVE_<MODEL_NAME>_SERVICE = service-name
    for _, env := range os.Environ() {
        if strings.HasPrefix(env, "KSERVE_") && strings.HasSuffix(env, "_SERVICE") {
            parts := strings.SplitN(env, "=", 2)
            if len(parts) == 2 {
                envKey := parts[0]
                serviceName := parts[1]

                // Convert KSERVE_DISK_FAILURE_PREDICTOR_SERVICE → disk-failure-predictor
                modelName := strings.TrimPrefix(envKey, "KSERVE_")
                modelName = strings.TrimSuffix(modelName, "_SERVICE")
                modelName = strings.ToLower(strings.ReplaceAll(modelName, "_", "-"))

                models[modelName] = serviceName
            }
        }
    }

    return models
}
```

This allows users to add custom models by updating `values-hub.yaml`:

```yaml
coordinationEngine:
  kserve:
    services:
      anomaly_detector: "anomaly-detector-predictor"
      predictive_analytics: "predictive-analytics-predictor"
      # USER ADDS CUSTOM MODEL:
      disk_failure_predictor: "disk-failure-predictor-predictor"
```

### 4. Health Checks for KServe Models

Implement health check endpoint to verify KServe model availability:

```
GET /api/v1/models/{model_name}/health
```

**Response**:
```json
{
  "model": "anomaly-detector",
  "status": "ready",
  "service": "anomaly-detector-predictor",
  "namespace": "self-healing-platform"
}
```

### File Structure

```
pkg/
├── kserve/
│   ├── client.go       # KServe HTTP client
│   └── models.go       # Model registry and discovery
├── api/
│   ├── detect.go       # /api/v1/detect handler
│   └── models.go       # /api/v1/models/:model/health handler
cmd/
└── server/
    └── main.go         # Register handlers
```

#### `pkg/kserve/client.go`

```go
package kserve

import (
    "bytes"
    "encoding/json"
    "fmt"
    "net/http"
    "os"
    "strings"
    "time"
)

type Client struct {
    httpClient *http.Client
    namespace  string
    models     map[string]string // model name -> service name
}

type PredictRequest struct {
    Instances [][]float64 `json:"instances"`
}

type PredictResponse struct {
    Predictions  []int  `json:"predictions"`
    ModelName    string `json:"model_name,omitempty"`
    ModelVersion string `json:"model_version,omitempty"`
}

func NewClient() (*Client, error) {
    namespace := os.Getenv("KSERVE_NAMESPACE")
    if namespace == "" {
        namespace = "self-healing-platform"
    }

    // Discover models from environment variables
    models := loadModelsFromEnv()

    if len(models) == 0 {
        return nil, fmt.Errorf("no KServe models configured")
    }

    return &Client{
        httpClient: &http.Client{Timeout: 30 * time.Second},
        namespace:  namespace,
        models:     models,
    }, nil
}

func loadModelsFromEnv() map[string]string {
    models := make(map[string]string)

    for _, env := range os.Environ() {
        if strings.HasPrefix(env, "KSERVE_") && strings.HasSuffix(env, "_SERVICE") {
            parts := strings.SplitN(env, "=", 2)
            if len(parts) == 2 {
                envKey := parts[0]
                serviceName := parts[1]

                modelName := strings.TrimPrefix(envKey, "KSERVE_")
                modelName = strings.TrimSuffix(modelName, "_SERVICE")
                modelName = strings.ToLower(strings.ReplaceAll(modelName, "_", "-"))

                models[modelName] = serviceName
            }
        }
    }

    return models
}

func (c *Client) Predict(modelName string, instances [][]float64) (*PredictResponse, error) {
    serviceName, ok := c.models[modelName]
    if !ok {
        return nil, fmt.Errorf("unknown model: %s", modelName)
    }

    // KServe v1 prediction API
    url := fmt.Sprintf("http://%s.%s.svc.cluster.local/v1/models/%s:predict",
        serviceName, c.namespace, modelName)

    reqBody := PredictRequest{Instances: instances}
    jsonData, err := json.Marshal(reqBody)
    if err != nil {
        return nil, err
    }

    resp, err := c.httpClient.Post(url, "application/json", bytes.NewBuffer(jsonData))
    if err != nil {
        return nil, fmt.Errorf("kserve request failed: %w", err)
    }
    defer resp.Body.Close()

    if resp.StatusCode != http.StatusOK {
        return nil, fmt.Errorf("kserve returned status: %d", resp.StatusCode)
    }

    var response PredictResponse
    if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
        return nil, fmt.Errorf("failed to decode response: %w", err)
    }

    return &response, nil
}

func (c *Client) HealthCheck(modelName string) error {
    serviceName, ok := c.models[modelName]
    if !ok {
        return fmt.Errorf("unknown model: %s", modelName)
    }

    url := fmt.Sprintf("http://%s.%s.svc.cluster.local/v1/models/%s",
        serviceName, c.namespace, modelName)

    resp, err := c.httpClient.Get(url)
    if err != nil {
        return fmt.Errorf("health check failed: %w", err)
    }
    defer resp.Body.Close()

    if resp.StatusCode != http.StatusOK {
        return fmt.Errorf("model not ready: %s (status: %d)", modelName, resp.StatusCode)
    }

    return nil
}

func (c *Client) ListModels() []string {
    models := make([]string, 0, len(c.models))
    for model := range c.models {
        models = append(models, model)
    }
    return models
}
```

#### `pkg/api/detect.go`

```go
package api

import (
    "encoding/json"
    "fmt"
    "log"
    "net/http"
    "coordination-engine/pkg/kserve"
)

type DetectRequest struct {
    Model     string      `json:"model"`
    Instances [][]float64 `json:"instances"`
}

type DetectHandler struct {
    kserveClient *kserve.Client
}

func NewDetectHandler(kserveClient *kserve.Client) *DetectHandler {
    return &DetectHandler{kserveClient: kserveClient}
}

func (h *DetectHandler) Handle(w http.ResponseWriter, r *http.Request) {
    if r.Method != http.MethodPost {
        http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
        return
    }

    var req DetectRequest
    if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
        log.Printf("Invalid request: %v", err)
        http.Error(w, "Invalid request format", http.StatusBadRequest)
        return
    }

    // Validate request
    if req.Model == "" {
        http.Error(w, "Missing 'model' field", http.StatusBadRequest)
        return
    }

    if len(req.Instances) == 0 {
        http.Error(w, "Missing 'instances' field", http.StatusBadRequest)
        return
    }

    // Call KServe model
    log.Printf("Calling KServe model: %s with %d instances", req.Model, len(req.Instances))
    resp, err := h.kserveClient.Predict(req.Model, req.Instances)
    if err != nil {
        log.Printf("Prediction failed: %v", err)
        http.Error(w, fmt.Sprintf("Prediction failed: %v", err), http.StatusInternalServerError)
        return
    }

    w.Header().Set("Content-Type", "application/json")
    json.NewEncoder(w).Encode(resp)
}
```

#### `pkg/api/models.go`

```go
package api

import (
    "encoding/json"
    "net/http"
    "coordination-engine/pkg/kserve"
)

type ModelHealthResponse struct {
    Model     string `json:"model"`
    Status    string `json:"status"`
    Service   string `json:"service"`
    Namespace string `json:"namespace"`
}

type ModelsHandler struct {
    kserveClient *kserve.Client
}

func NewModelsHandler(kserveClient *kserve.Client) *ModelsHandler {
    return &ModelsHandler{kserveClient: kserveClient}
}

func (h *ModelsHandler) ListModels(w http.ResponseWriter, r *http.Request) {
    models := h.kserveClient.ListModels()

    w.Header().Set("Content-Type", "application/json")
    json.NewEncoder(w).Encode(map[string]interface{}{
        "models": models,
        "count":  len(models),
    })
}
```

#### `cmd/server/main.go` Updates

```go
// Add to main() function after existing handlers

// Initialize KServe client if enabled
if os.Getenv("ENABLE_KSERVE_INTEGRATION") == "true" {
    kserveClient, err := kserve.NewClient()
    if err != nil {
        log.Printf("⚠️  KServe integration disabled: %v", err)
    } else {
        log.Println("✅ KServe integration enabled")
        log.Printf("📊 Registered models: %v", kserveClient.ListModels())

        // Register detection endpoint
        detectHandler := api.NewDetectHandler(kserveClient)
        mux.HandleFunc("/api/v1/detect", detectHandler.Handle)
        log.Println("✅ Registered /api/v1/detect endpoint")

        // Register models list endpoint
        modelsHandler := api.NewModelsHandler(kserveClient)
        mux.HandleFunc("/api/v1/models", modelsHandler.ListModels)
        log.Println("✅ Registered /api/v1/models endpoint")
    }
}
```

### POST /api/v1/detect

Call KServe model for predictions.

**Request**:
```json
{
  "model": "anomaly-detector",
  "instances": [[0.5, 1.2, 0.8]]
}
```

**Response** (200 OK):
```json
{
  "predictions": [-1],
  "model_name": "anomaly-detector",
  "model_version": "v2"
}
```

**Errors**:
- `400 Bad Request`: Invalid request format
- `404 Not Found`: Model not registered
- `500 Internal Server Error`: KServe unavailable or prediction failed

### GET /api/v1/models

List all registered KServe models.

**Response**:
```json
{
  "models": [
    "anomaly-detector",
    "predictive-analytics",
    "disk-failure-predictor"
  ],
  "count": 3
}
```

### Unit Tests

```go
// pkg/kserve/client_test.go
func TestLoadModelsFromEnv(t *testing.T) {
    os.Setenv("KSERVE_ANOMALY_DETECTOR_SERVICE", "anomaly-detector-predictor")
    os.Setenv("KSERVE_DISK_FAILURE_PREDICTOR_SERVICE", "disk-failure-predictor-predictor")

    models := loadModelsFromEnv()

    assert.Equal(t, "anomaly-detector-predictor", models["anomaly-detector"])
    assert.Equal(t, "disk-failure-predictor-predictor", models["disk-failure-predictor"])
}
```

### Integration Tests

```bash
# Test coordination engine KServe proxy
oc exec deployment/coordination-engine -n self-healing-platform -- \
  curl -X POST http://localhost:8080/api/v1/detect \
  -H "Content-Type: application/json" \
  -d '{"model": "anomaly-detector", "instances": [[0.5, 1.2, 0.8]]}'

# Expected: {"predictions": [-1], "model_name": "anomaly-detector"}

# Test model listing
oc exec deployment/coordination-engine -n self-healing-platform -- \
  curl http://localhost:8080/api/v1/models

# Expected: {"models": ["anomaly-detector", "predictive-analytics"], "count": 2}
```

## User Workflow: Adding Custom Models

This implementation enables a seamless workflow for users to add custom models:

### 1. Train Model in Notebook

```python
from sklearn.ensemble import IsolationForest
import joblib

model = IsolationForest(contamination=0.1)
model.fit(X_train)
joblib.dump(model, '/opt/app-root/src/models/disk-failure-predictor/model.pkl')
```

### 2. Deploy as KServe InferenceService

```yaml
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: disk-failure-predictor
  namespace: self-healing-platform
spec:
  predictor:
    model:
      modelFormat:
        name: sklearn
      storageUri: "pvc://model-storage-pvc/disk-failure-predictor"
```

### 3. Register in values-hub.yaml

```yaml
coordinationEngine:
  kserve:
    services:
      disk_failure_predictor: "disk-failure-predictor-predictor"
```

### 4. Use from Notebook

```python
result = coord_client.detect_anomaly(
    model_name="disk-failure-predictor",
    instances=[[85.5, 5000, 365]]
)
```

**No coordination engine code changes needed!** ✅

## Configuration (Already Complete)

The platform deployment is already configured:

**`values-hub.yaml`**:
```yaml
coordinationEngine:
  kserve:
    enabled: true
    namespace: self-healing-platform
    services:
      anomaly_detector: "anomaly-detector-predictor"
      predictive_analytics: "predictive-analytics-predictor"
```

**`charts/hub/templates/coordination-engine.yaml`**:
```yaml
env:
  - name: ENABLE_KSERVE_INTEGRATION
    value: {{ .Values.coordinationEngine.kserve.enabled | quote }}
  - name: KSERVE_NAMESPACE
    value: {{ .Values.coordinationEngine.kserve.namespace | quote }}
  {{- range $key, $value := .Values.coordinationEngine.kserve.services }}
  - name: KSERVE_{{ $key | upper }}_SERVICE
    value: {{ $value | quote }}
  {{- end }}
```

## References

- **ADR-004**: [KServe for Model Serving Infrastructure](https://github.com/tosin2013/openshift-aiops-platform/blob/main/docs/adrs/004-kserve-model-serving.md)
- **ADR-038**: [Go Coordination Engine Migration](https://github.com/tosin2013/openshift-aiops-platform/blob/main/docs/adrs/038-go-coordination-engine-migration.md)
- **ADR-039**: [User-Deployed KServe Models](https://github.com/tosin2013/openshift-aiops-platform/blob/main/docs/adrs/039-user-deployed-kserve-models.md)
- **ADR-040**: [Extensible KServe Model Registry](https://github.com/tosin2013/openshift-aiops-platform/blob/main/docs/adrs/040-extensible-kserve-model-registry.md)
- **User Guide**: [USER-MODEL-DEPLOYMENT-GUIDE.md](https://github.com/tosin2013/openshift-aiops-platform/blob/main/docs/guides/USER-MODEL-DEPLOYMENT-GUIDE.md)
- **KServe v1 Protocol**: https://kserve.github.io/website/latest/modelserving/data_plane/v1_protocol/

## Success Criteria

- [ ] `/api/v1/detect` endpoint implemented
- [ ] Dynamic model discovery from environment variables
- [ ] `/api/v1/models` endpoint lists registered models
- [ ] KServe v1 protocol support
- [ ] Integration tests pass
- [ ] Users can add custom models via values-hub.yaml
- [ ] Notebook updated to use coordination engine proxy (see issue #15)

## Related Issues

- **Issue #15**: Update coordination-engine-integration.ipynb to use coordination engine proxy
- **Issue #16**: Update rule-based-remediation.ipynb API endpoint (already fixed)

## Labels

`enhancement`, `kserve`, `adr-039`, `adr-040`, `ml-integration`, `coordination-engine`
